### ISSUE TITLE
Support for macro imports in parent templates

### ISSUE THREAD

**Issue by edwardhorsford** [11ty/eleventy#212]
I'm coming from a node / nunjucks app where I import my macros once in my parent layout template like so:

`{% import "includes/macros.html" as macro with context %}`

Now all pages that use this template get access to my macros.

I couldn't get this to work with eleventy - it *appears* like every page needs to reference the macros file individually. It would be great if they could be referenced just the once.

Some conversation on twitter about this:
https://twitter.com/adambsilver/status/1038330552363966464

**Comment by zachleat** [comment]
Is your node/nunjucks app using Nunjucks `extends` for layouts? We do still support those.

Hmm, this one actually might be difficult to get working with eleventy layouts. Structurally, we render layout-less output of templates for `templateContent` strings in collection entries.
https://www.11ty.io/docs/collections/#collection-item-data-structure

**Comment by edwardhorsford** [comment]
No, I'm using the Eleventyone project as a starter. I can give extends a try - that's what my existing site uses.

Is the an advantage / disadvantage to using one or the other?

**Comment by zachleat** [comment]
If you use Nunjucks extends for layouts, you can’t use front matter in your layout to set inherited data. But you can use directory data files to do this same thing https://www.11ty.io/docs/data-template-dir/

The other thing Eleventy layouts gives you is cross-template layout use. Eleventy layouts can use any template language, independent of the child content. This may or may not be important for your project.

**Comment by edwardhorsford** [comment]
I've tried using extends in my layouts and I *think* it still doesn't work when importing macros in the parent template.

**Comment by zachleat** [comment]
Sounds like this is going to need a new feature to support this request. Possibly a front matter convention—not sure yet? I’m open to ideas!

**Comment by edwardhorsford** [comment]
I've ended up including the import on every page, including templates. It's not a major thing, but a bit annoying.

Thankfully eleventy errors obviously well when it's *not* included, so I know pretty quickly when I've messed it up.

Of course my ideal would be that each page only has what's *unique* about that page - with imports and such coming from higher level templates. But I'm not sure it's worth adding more magic to frontmatter to achieve this.

**Comment by paulmsmith** [comment]
+1 To this request.

I've tried a few tricks to get this working but was fighting Eleventy layouts. I tried creating a nunjucks filter that exposed the `addGlobal` method but it didn't work, probably with how eleventy preps the templates for render.

**Comment by mashdot** [comment]
Glad I finally found this thread. Spent over an hour trying to get my macros included via templates layouts. Add the import each page manually will be a pain.

**Comment by aharvard** [comment]
I agree! Being able to import my macros in the parent template would be extremely helpful. 

@zachleat, I think a front matter convention could work. I'm not sure how this would look for other template languages, but for nunjucks maybe something like? 

```
---
layout: parent.njk
eleventyImport:
  file: macros.njk
  as: macro
  context: true
---
```

for multiple imports:
```
---
layout: parent.njk
eleventyImport:
  - file: macros.njk
    as: macro
    context: true
  - file: my-components.njk
    as: components
---
```

**Comment by aharvard** [comment]
In the meantime, like @edwardhorsford, I'm just including the import on every page. 

Not a huge deal. Just a nice to have. Thanks for starting the convo!

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by erikwoods** [comment]
I also spent hours trying to find things online to help with this.

I tried using Nunjucks `extends` and experienced the same issue. Child templates don't seem to have access to or knowledge of macros imported by parent templates.

Do I really still have to import all of my macros on every template that I want to use them on? Or has there been a solution implemented for this that I just haven't found in the docs yet?

One of my goals with this app I'm building is to eliminate barriers for people to contribute - and my typical user won't want to be bothered with this sort of thing.

**Comment by edwardhorsford** [comment]
@erikwoods I looked at this again last week and still couldn't find a solution. I'd love it if eleventy could do something. Prepend content maybe?

At the moment, I do two things:
* I keep macros split across several files, but in my build pipeline, concat them in to a single `macros.njk` file. This then gets imported once in each template (my templates have the line already included).
* Depending on how your macros are set up, you might be able to make a single file that creates the macros, but imports the actual content from other files, and then import that - see example below.

Eg in uk gov ([example here](https://github.com/alphagov/govuk-frontend/tree/master/src/govuk/components/button)), the html for the macro is kept separate from the creation of it.

template.njk:
```
// whatever markup you want to output
```
macro.njk:
```njk
{% macro govukButton(params) %}
  {%- include "./template.njk" -%}
{% endmacro %}
```

What I've I realised last week (when wanting to import all UK gov macros) is that the above format allows me to have one file like:

extraMacros.njk:
```njk
{% macro firstComponent(params) %}
  {%- include "./firstComponent/template.njk" -%}
{% endmacro %}

{% macro secondComponent(params) %}
  {%- include "./secondComponent/template.njk" -%}
{% endmacro %}
```

This file then gets concatted with my main macros so that they all become available under the `macro` namespace.

Not ideal, but means you wouldn't have to import _each_ on every single page.

**Comment by erikwoods** [comment]
Thanks for the suggestion @edwardhorsford !
I will have to check it out ASAP.

**Comment by aaronstezycki** [comment]
@edwardhorsford I've tried this technique out, and it does seem to work... only I'd imagine theres alot of setup and code repetition involved. 

I've got a few call tags used within the components im using, for extra flexibility with content, is there a way this technique could be used for those instances?

**Comment by zachleat** [comment]
A path forward for this is available at #3345 and #188

**Comment by erikwoods** [comment]
> A path forward for this is available at #3345 and #188

Wow - thanks! Haven't tested it out yet but seems promising! Especially since it's coming from you LOL. PS I'd like to connect on LinkedIn - sent you a request.
